### PR TITLE
added: addon callback for listing a directory

### PR DESCRIPTION
**Pull request by notspiff**
This give add-ons the opportunity to list vfs paths. Code is extracted from my vfs addon work. I have not actually tested this incarnation but it should work just fine (i have tested the _code_ extensively).

Likely intefers with your work @AlwinEsch, i can rebase it after your stuff lands if you prefer no problem.

### CODE REVIEW

File: xbmc/addons/include/kodi_vfs_types.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  struct VFSDirEntry
+  {
+    char* label;             //!< item label
+    char* title;             //!< item title
+    char* path;              //!< item path
+    int num_props;           //!< Number of properties attached to item
+    VFSProperty* properties; //!< Properties
+    bool folder;             //!< Item is a folder
+    uint64_t size;           //!< Size of file represented by item
+  };
```
Comment (AlwinEsch) [inline xbmc/addons/include/kodi_vfs_types.h]: There are few values e.g. title not set on kodi. Is this planned for add-on side?

Comment (akva2) [inline xbmc/addons/include/kodi_vfs_types.h]: yeah, that would be a "remnant" from the vfs addons. they use this struct to transfer information to kodi.

File: addons/library.xbmc.addon/libXBMC_addon.h [hunk 2]
```diff
@@ -264,6 +266,14 @@ namespace ADDON
...
       if (XBMC_remove_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
 
+      XBMC_get_directory = (bool (*)(void* HANDLE, void* CB, const char* strPath, const char* mask, VFSDirEntry** items, int* num_items))
+        dlsym(m_libXBMC_addon, "XBMC_get_directory");
+      if (XBMC_get_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
+
+      XBMC_free_directory = (void (*)(void* HANDLE, void* CB, VFSDirEntry* items, int num_items))
+        dlsym(m_libXBMC_addon, "XBMC_free_directory");
+      if (XBMC_free_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
+
```
Comment (AlwinEsch) [inline addons/library.xbmc.addon/libXBMC_addon.h]: One point for the library rework is that also dlsym is no more needed, except the "Register" and "Unregister".

There comes also a question (idea). Possible to use then ```std::vector``` and ```std::string``` direct and prevent the use of ```free``` ?

Comment (akva2) [inline addons/library.xbmc.addon/libXBMC_addon.h]: If the policy to keep the api c has changed it is an option. Should be fairly hassle less with the c++11 abi standardization.

-------- Opprinnelig melding --------
Fra: Alwin Esch <notifications@github.com>
Dato: 27.11.2015 18:03 (GMT+01:00)
Til: xbmc/xbmc <xbmc@noreply.github.com>
Ko: Arne Morten Kvarving <Arne.Morten.Kvarving@sintef.no>
Emne: Re: [xbmc] added: addon callback for listing a directory (#8450)


In addons/library.xbmc.addon/libXBMC_addon.h<https://github.com/xbmc/xbmc/pull/8450#discussion_r46061875>:

> @@ -264,6 +266,14 @@ namespace ADDON
>          dlsym(m_libXBMC_addon, "XBMC_remove_directory");
>        if (XBMC_remove_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
>
> +      XBMC_get_directory = (bool (*)(void* HANDLE, void* CB, const char* strPath, const char* mask, VFSDirEntry** items, int* num_items))
> +        dlsym(m_libXBMC_addon, "XBMC_get_directory");
> +      if (XBMC_get_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
> +
> +      XBMC_free_directory = (void (*)(void* HANDLE, void* CB, VFSDirEntry* items, int num_items))
> +        dlsym(m_libXBMC_addon, "XBMC_free_directory");
> +      if (XBMC_free_directory == NULL) { fprintf(stderr, "Unable to assign function %s\n", dlerror()); return false; }
> +


One point for the library rework is that also dlsym is no more needed, except the "Register" and "Unregister".

There comes also a question (idea). Possible to use then std::vector and std::string direct and prevent the use of free ?

File: xbmc/addons/include/kodi_vfs_types.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+extern "C"
+{
+  struct VFSDirEntry
+  {
+    char* label;             //!< item label
+    char* path;              //!< item path
+    bool folder;             //!< Item is a folder
+    uint64_t size;           //!< Size of file represented by item
+  };
+}
```
Comment (AlwinEsch) [inline xbmc/addons/include/kodi_vfs_types.h]: Ok for you to this on end in related library header e.g.
https://github.com/AlwinEsch/kodi/blob/binary-lib-rework-2/addons/library.kodi.addon/addon.api2/definations.h
or
https://github.com/AlwinEsch/kodi/blob/binary-lib-rework-2/addons/library.kodi.addon/addon.api2/Directory.h
?

Comment (notspiff) [inline xbmc/addons/include/kodi_vfs_types.h]: apart from the 'definations' name (it's definItions) it seems very nice.

**Comment by AlwinEsch** [comment]
On my rework is the most done, here (at home) are all library parts finished and working now on cleanup of them. The biggest part to do on end is to add a documentation for them, but I hope to have finished the next days.

How it used is visible here https://github.com/AlwinEsch/pvr.vdr.vnsi/commit/567b580541face2abb7d8b4bf195d915cc7d25a6

**Comment by garbear** [comment]
I have a solution to the file properties window. Github's not letting me PR to you, but you can cherry-pick my addition here: https://github.com/garbear/xbmc/compare/cbc72f3...add_directory_listing_callback

I added a new file, kodi_vfs_utils.hpp, similar to xbmc_stream_utils.hpp. It's a raw C++ wrapper around the C api. So we get the best of both worlds. C-compatible struct arrays, and C++ compatible vectors.

the util is called like this:

```c++
ADDON::VFSUtils:: GetDirectory(XBMC, path, mask, items);
```

**Comment by AlwinEsch** [comment]
Has reworked and finished alpha of GUI library in new design. Everything is handled with std::string now, here with test on linux (not complete now, only with pvr.vdr.vnsi tested) was everything OK. 

https://github.com/AlwinEsch/kodi/tree/binary-lib-rework-2/addons/library.kodi.guilib/gui.api2

Except a few calls ~4-5 (few are inline) is no more special work needed to call work on Kodi.

Now I rework the others and then comes the documentation.
Is this way about library OK for you? Is a lot of more files but for me better for programming from others on add-on and for me to find the place where I want to do (before everything in one file was to much for me).

**Comment by AlwinEsch** [comment]
Thanks @garbear, is really good usable and becomes included.

What do you think about the use of std::string in guilib, it has pro and contra for me.

Pro for me is:
- No special reworks of data from c++ to c and back to c++
- I think better performance
- The risk for leaks is not so high
- Maybe better exception handling

Contra:
- Not usable over the dlopen way if it comes back for some reason
- Correct work with C++ from me only confirmed on linux, Windows still to do.
- Is it really compatible if compiled from different compiler versions (????)
- How it becomes used on other programming languages (with web engine comes also in future all his systems to kodi e.g. Java-Script or Perl)

**Comment by garbear** [comment]
the best thing to do is stick with a C api. that solves the compatibility issue. Next, the C api should be wrapped in C++. The wrapper should be as thin as possible, just allocate/free memory and other C++ -> C conversions.

For example, check out my C++ wrapper for the peripheral API: [kodi_peripheral_utils.hpp](https://github.com/garbear/xbmc/blob/retroplayer-15.2/xbmc/addons/include/kodi_peripheral_utils.hpp). Each struct in the kodi_peripheral_types.h file is mirrored by a C++ class that handles the memory.

Now, my add-on deals exclusively with C++ classes. Right before I pass the struct across the DLL boundary, I serialize to a C struct. On the other side (in kodi application) I change back to C++ immediately. The only time a C struct is exposed to my code is the single invocation of the callback.

Make sure that the C++ wrappers are as thin as possible. Thin layers built on top of each other is a good strategy because it maximizes the use of abstraction.

**Comment by AlwinEsch** [comment]
Hi @garbear and @notspiff, have added your listing a directory parts here https://github.com/AlwinEsch/kodi/commit/b7312bf47dc14cf1ed976dc4ab9ed5ca865847ce
to the new version of library.

Hope is OK for you.

**Comment by AlwinEsch** [comment]
Has forgot on last commit to add two files.
https://github.com/AlwinEsch/kodi/commit/f72324b3ce855b6ad3163bdcff8ab4860e0c4548

**Comment by garbear** [comment]
@AlwinEsch looks good

BTW I've added two binary add-on APIs in my RetroPlayer branch: the Game API and the Peripheral API. The headers are here: https://github.com/garbear/xbmc/tree/retroplayer-15.2/xbmc/addons/include. The API's aren't done yet, but I'm trying to finish the peripheral API by january. just a heads up, we'll need to adapt those ones too

**Comment by akva2** [comment]
I dont think you will be very happy about these changes garrett. None of your preload hacks will work when there is no dynloading of the callbacks.

**Comment by AlwinEsch** [comment]
@akva2 what you mean with preload hacks? Can you give me a link to see it.

@garbear has seen you use also few general thinks like this https://github.com/garbear/xbmc/blob/retroplayer-15.2/xbmc/addons/include/kodi_addon_utils.hpp#L108

I add it direct to my changes. If anyone else have general things like this please inform me.

**Comment by garbear** [comment]
@AlwinEsch Have you added RIAA wrappers for strings like the ones in kodi_addon_utils.hpp? best to hide all the memory management from the client

**Comment by akva2** [comment]
I havent looked into the code in a while but basically there are/were multiple implementations of the callback classes and a particular one was chosen by preloading the appropriate .so.

-------- Opprinnelig melding --------
Fra: Alwin Esch <notifications@github.com>
Dato: 29.11.2015 18:36 (GMT+01:00)
Til: xbmc/xbmc <xbmc@noreply.github.com>
Ko: Arne Morten Kvarving <Arne.Morten.Kvarving@sintef.no>
Emne: Re: [xbmc] added: addon callback for listing a directory (#8450)


@akva2<https://github.com/akva2> what you mean with preload hacks? Can you give me a link to see it.

@garbear<https://github.com/garbear> has seen you use also few general thinks like this https://github.com/garbear/xbmc/blob/retroplayer-15.2/xbmc/addons/include/kodi_addon_utils.hpp#L108

I add it direct to my changes. If anyone else have general things like this please inform me.

**Comment by AlwinEsch** [comment]
@garbear what is RIAA? There becomes the memory managment handled from the headers itself and returns always e.g. a std::string.

It use currently two different ways. One is where the Kodi gives a allocated memory where add-on header calls free on end. The other (for smaller strings) is where the add-on send a c translated std::string memory array pointer to Kodi with given max size where kodi copies the data to it.

**Comment by garbear** [comment]
lol I meant RAII. "Resource Acquisition Is Initialization". you encapsulate all allocations and deallocations in the same place.

You do this in [libKODI_addon_Internal.h](https://github.com/AlwinEsch/kodi/blob/binary-lib-rework-2/addons/library.kodi.addon/addon.api2/internal/libKODI_addon_Internal.h#L246). It looks like there's a few missing null checks (e.g. GetLocalizedString()) but otherwise returning a `std::string` instead of char array that needs to be freed is a good thing.

**Comment by garbear** [comment]
@akva2 still not sure what you mean. are you talking about the libretro wrapper loading libretro DLLs?

**Comment by notspiff** [comment]
i thought the network kode replaced libKODI_game but i was wrong.

**Comment by AlwinEsch** [comment]
Hi @garbear, has leaved it away and thought has already present checks inside related ```std::string``` and ```free``` does only the work if the pointer is not NULL.

Has checked it and brought exception with NULL. But the good thing was again that ```ERROR: AddOnLog: VDR VNSI Client: CallMenuHook - std::exception``` has worked. Juhu.

Can be this come due to direct calls on a function table instead of the use of ```dlsym``` on library?

**Comment by garbear** [comment]
@AlwinEsch are you talking about your recent fix for null checks? that looks good and fixes the example I mentioned, thanks

Passing NULL should throw an exception because you're constructing a std::string from NULL, e.g.

```c++
std::string str = NULL; // error
```

Your english isn't very clear, so I'm not sure if this answers your question.

**Comment by garbear** [comment]
@akva2 I sent a PR your way containing a couple fixes I've found

**Comment by Razzeee** [comment]
What's the status here?

**Comment by notspiff** [comment]
Included in garret's input pr i believe.
8. jan. 2016 18:32 skrev "Razzeee" <notifications@github.com>:

**Comment by garbear** [comment]
doesn't look like it

**Comment by garbear** [comment]
wait - nvm - can we merge this first, with my fixes, so I can rebase this PR out of mine?

**Comment by garbear** [comment]
@notspiff https://github.com/notspiff/xbmc/pull/2

**Comment by notspiff** [comment]
pull in and squashed (it was all fixes and we don't want to pull in broken revisions).

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by garbear** [comment]
ready to go, yea?

**Comment by garbear** [comment]
jenkins build and merge

**Pull request merged**